File: src-tauri/src/widgets/toggles.rs
//! 시스템 토글: 다크모드 / Wi-Fi / Bluetooth / 음소거 / 잠금방지(caffeinate)
//! 모든 상태조회/변경은 wifi 장치명과 caffeinate 상태를 파라미터로 받는다 (ipc.rs에서 주입)

use serde::Serialize;

#[derive(Serialize)]
pub struct TogglesState {
    pub dark: bool,
    pub wifi: bool,
    pub bt: bool,
    pub mute: bool,
    pub caffeine: bool,
    pub wifi_available: bool,
    pub bt_available: bool,
}

pub fn detect_wifi_device() -> Option<String> {
    let out = crate::platform::shell_output("networksetup", &["-listallhardwareports"])?;
    let mut lines = out.lines();
    while let Some(line) = lines.next() {
        if line.trim() == "Hardware Port: Wi-Fi" {
            if let Some(dev_line) = lines.next() {
                if let Some(dev) = dev_line.strip_prefix("Device: ") {
                    return Some(dev.trim().to_string());
                }
            }
        }
    }
    None
}

pub fn get_state(wifi_dev: &Option<String>, caffeine_on: bool) -> TogglesState {
    let dark = crate::platform::shell_output("defaults", &["read", "-g", "AppleInterfaceStyle"])
        .map(|v| v.eq_ignore_ascii_case("Dark"))
        .unwrap_or(false);

    let wifi = wifi_state(wifi_dev).unwrap_or(false);

    let bt_available = crate::platform::command_exists("blueutil");
    let bt = bt_available
        .then(|| crate::platform::shell_output("blueutil", &["-p"]))
        .flatten()
        .map(|v| v == "1")
        .unwrap_or(false);

    let mute = crate::platform::osascript("output muted of (get volume settings)")
        .map(|v| v == "true")
        .unwrap_or(false);

    TogglesState {
        dark,
        wifi,
        bt,
        mute,
        caffeine: caffeine_on,
        wifi_available: wifi_dev.is_some(),
        bt_available,
    }
}

pub fn wifi_state(dev: &Option<String>) -> Option<bool> {
    let dev = dev.as_ref()?;
    let out = crate::platform::shell_output("networksetup", &["-getairportpower", dev])?;
    Some(out.to_lowercase().contains("on"))
}

pub fn wifi_set(dev: &Option<String>, on: bool) -> Result<(), String> {
    let Some(dev) = dev else {
        return Err("Wi-Fi 장치를 찾을 수 없습니다".into());
    };
    let state = if on { "on" } else { "off" };
    crate::platform::shell_output("networksetup", &["-setairportpower", dev, state])
        .ok_or_else(|| "networksetup 실패 (권한 필요)".to_string())?;
    Ok(())
}

pub fn dark_mode_set(on: bool) -> Result<(), String> {
    let val = if on { "true" } else { "false" };
    crate::platform::osascript(&format!(
        "tell application \"System Events\" to tell appearance preferences to set dark mode to {val}"
    ))
    .map(|_| ())
}

pub fn mute_set(on: bool) -> Result<(), String> {
    let val = if on { "true" } else { "false" };
    crate::platform::osascript(&format!("set volume output muted {val}")).map(|_| ())
}

pub fn bt_set(on: bool) -> Result<(), String> {
    if !crate::platform::command_exists("blueutil") {
        return Err("blueutil이 필요합니다: brew install blueutil".into());
    }
    let val = if on { "1" } else { "0" };
    crate::platform::shell_output("blueutil", &["-p", val])
        .ok_or_else(|| "blueutil 실행 실패".to_string())?;
    Ok(())
}

File: goport/toggles.go
package main

// 시스템 토글: 다크모드 / Wi-Fi / Bluetooth / 음소거 / 잠금방지(caffeinate)
// (Rust widgets/toggles.rs 포트)

import (
	"fmt"
	"os/exec"
	"strings"
)

type TogglesState struct {
	Dark          bool `json:"dark"`
	Wifi          bool `json:"wifi"`
	Bt            bool `json:"bt"`
	Mute          bool `json:"mute"`
	Caffeine      bool `json:"caffeine"`
	WifiAvailable bool `json:"wifi_available"`
	BtAvailable   bool `json:"bt_available"`
}

func detectWifiDevice() string {
	out, ok := shellOutput("networksetup", "-listallhardwareports")
	if !ok {
		return ""
	}
	lines := strings.Split(out, "\n")
	for i, line := range lines {
		if strings.TrimSpace(line) == "Hardware Port: Wi-Fi" {
			for _, devLine := range lines[i+1:] {
				if strings.HasPrefix(devLine, "Device: ") {
					return strings.TrimSpace(strings.TrimPrefix(devLine, "Device: "))
				}
			}
		}
	}
	return ""
}

func wifiState(dev string) (bool, bool) {
	if dev == "" {
		return false, false
	}
	out, ok := shellOutput("networksetup", "-getairportpower", dev)
	if !ok {
		return false, false
	}
	return strings.Contains(strings.ToLower(out), "on"), true
}

func getTogglesState(st *AppState) TogglesState {
	dark := false
	if out, ok := shellOutput("defaults", "read", "-g", "AppleInterfaceStyle"); ok {
		dark = strings.EqualFold(out, "Dark")
	}
	wifi, _ := wifiState(st.wifiDev)
	btAvailable := commandExists("blueutil")
	bt := false
	if btAvailable {
		if out, ok := shellOutput("blueutil", "-p"); ok {
			bt = out == "1"
		}
	}
	mute := false
	if out, err := osascript("output muted of (get volume settings)"); err == nil {
		mute = out == "true"
	}
	caffeinate := false
	st.caffeinateMu.Lock()
	caffeinate = st.caffeinate != nil
	st.caffeinateMu.Unlock()
	return TogglesState{
		Dark:          dark,
		Wifi:          wifi,
		Bt:            bt,
		Mute:          mute,
		Caffeine:      caffeinate,
		WifiAvailable: st.wifiDev != "",
		BtAvailable:   btAvailable,
	}
}

func wifiSet(dev string, on bool) error {
	if dev == "" {
		return fmt.Errorf("Wi-Fi 장치를 찾을 수 없습니다")
	}
	state := "off"
	if on {
		state = "on"
	}
	if _, ok := shellOutput("networksetup", "-setairportpower", dev, state); !ok {
		return fmt.Errorf("networksetup 실패 (권한 필요)")
	}
	return nil
}

func darkModeSet(on bool) error {
	val := "false"
	if on {
		val = "true"
	}
	_, err := osascript(`tell application "System Events" to tell appearance preferences to set dark mode to ` + val)
	return err
}

func muteSet(on bool) error {
	val := "false"
	if on {
		val = "true"
	}
	_, err := osascript("set volume output muted " + val)
	return err
}

func btSet(on bool) error {
	if !commandExists("blueutil") {
		return fmt.Errorf("blueutil이 필요합니다: brew install blueutil")
	}
	val := "0"
	if on {
		val = "1"
	}
	if _, ok := shellOutput("blueutil", "-p", val); !ok {
		return fmt.Errorf("blueutil 실행 실패")
	}
	return nil
}

func caffeineToggle(st *AppState) (bool, error) {
	st.caffeinateMu.Lock()
	defer st.caffeinateMu.Unlock()
	if st.caffeinate != nil {
		if st.caffeinate.Process != nil {
			_ = st.caffeinate.Process.Kill()
		}
		st.caffeinate = nil
		return false, nil
	}
	cmd := exec.Command("caffeinate", "-d")
	if err := cmd.Start(); err != nil {
		return false, fmt.Errorf("caffeinate 실행 실패: %v", err)
	}
	st.caffeinate = cmd
	go func() {
		_ = cmd.Wait()
		st.caffeinateMu.Lock()
		if st.caffeinate == cmd {
			st.caffeinate = nil
		}
		st.caffeinateMu.Unlock()
	}()
	return true, nil
}
